runtime/src/bank.rs

```diff
@@ -375,33 +375,38 @@ impl CachedExecutors {
         })
     }
 
-    fn put(&mut self, pubkey: &Pubkey, executor: Arc<dyn Executor>) {
-        let entry = if let Some(mut entry) = self.executors.remove(pubkey) {
-            entry.executor = executor;
-            entry
-        } else {
-            if self.executors.len() >= self.max {
-                let mut least = u64::MAX;
-                let default_key = Pubkey::default();
-                let mut least_key = &default_key;
-
-                for (key, entry) in self.executors.iter() {
-                    let count = entry.prev_epoch_count + entry.epoch_count.load(Relaxed);
-                    if count < least {
-                        least = count;
-                        least_key = key;
-                    }
-                }
-                let least_key = *least_key;
-                let _ = self.executors.remove(&least_key);
+    fn put(&mut self, pubkey: Pubkey, executor: Arc<dyn Executor>) {
+        match self.executors.entry(pubkey) {
+            Entry::Vacant(entry) => {
+                entry.insert(CachedExecutorsEntry {
+                    prev_epoch_count: u64::default(),
+                    epoch_count: AtomicU64::default(),
+                    executor,
+                });
             }
-            CachedExecutorsEntry {
-                prev_epoch_count: 0,
-                epoch_count: AtomicU64::new(0),
-                executor,
+            Entry::Occupied(mut entry) => {
+                entry.get_mut().executor = executor;
             }
         };
-        let _ = self.executors.insert(*pubkey, entry);
+        // The cache is allowed to grow upto 2 * capacity in size. Once there,
+        // extra entries are evicted incurring a linear computation cost. Since
+        // it takes at least O(n) calls to self.put(...) to invoke another
+        // eviction, the amoritzed cost is constant.
+        if self.executors.len() > self.max.saturating_mul(2) {
```

Comment:
The current max cached executors is 100 and we are now seeing programs around the 1MB, doubling the cache size is significant

Reply:
Times 2 is arbitrary. Any `x > 1`  will do. It will be still O(1) but with a different constant. 
e.g. We can change this to allow 50% overshoot before evicting extra elements:
```rust
if 100 * self.executors.len() > 150 * self.max {
```

What is the largest amount of overshoot that you consider feasible here?

Reply:
Yup, I like the method, not sure about the tuning.  @t-nelson do you have any info on how much memory nodes can afford to give to the executor cache?

Reply:
Would max programs per transaction (34) be sufficient overflow?

Reply:
If we don't keep the overflow executors then what is the purpose of the overflow, my understanding is the overflow needs to sufficiently large enough to allow for programs to graduate out of it.

Reply:
> Do we actually need to keep the overflow executors around or just their counts?

Yes, we need to keep the overflow executors.

I mentioned in the chat another idea about keeping counts without the actual executors:
> maintain `{,prev_}epoch_count` for 1000 pubkeys, but actually cache the executor for only the top 100.
> because maintaining the counts is cheap, but it helps with not thrashing the newest inserted item.
> the goal being when you actually cache the executor, you have a better estimate of the counts than just zeros

but that is an idea separate from this commit.

Reply:
https://github.com/solana-labs/solana/pull/25600 increased the cache size by 2.5x, whereas this code was just 2x at worst-case (and then even reduced to 1.7x).